### PR TITLE
CXX-692: findAndModify take a write concern

### PR DESCRIPTION
**Pull request by markbenvenuto**
Add a `writeConcern` argument to `findAndModify`.

I needed to move `findAndModify` to `DBClientBase` since this is the class that had `writeConcern` has the connection-level default `writeConcern`.

### CODE REVIEW

File: src/mongo/integration/standalone/dbclient_test.cpp [hunk 1]
```diff
@@ -240,6 +240,23 @@ TEST_F(DBClientTest, FindAndModifyDuplicateKeyError) {
...
+                                      BSON("$inc" << BSON("i" << 1)),
+                                      false,
+                                      false,
+                                      BSONObj(),
+                                      BSONObj(),
+                                      &WriteConcern::journaled);
+
+    ASSERT_EQUALS(result.getIntField("_id"), 1);
+    ASSERT_EQUALS(result.getIntField("i"), 1);
+    ASSERT_EQUALS(c->count(TEST_NS), 1U);
```
Comment (amidvidy) [inline src/mongo/integration/standalone/dbclient_test.cpp]: is it worth doing the whole song & dance with the system profiler to check that the system actually registered the proper write concern?

File: src/mongo/client/dbclient.cpp [hunk 2]
```diff
@@ -1365,6 +1366,13 @@ void DBClientWithCommands::_findAndModify(const StringData& ns,
     commandBuilder.append("new", returnNew);
     commandBuilder.append("upsert", upsert);
 
+    if (getMaxWireVersion() >= 4) {
+        const WriteConcern* operationWriteConcern =
+            writeConcern ? writeConcern : &getWriteConcern();
```
Comment (amidvidy) [inline src/mongo/client/dbclient.cpp]: will &getWriteConcern() always be non-null?

File: src/mongo/client/dbclient.cpp [hunk 3]
```diff
@@ -1365,6 +1366,13 @@ void DBClientWithCommands::_findAndModify(const StringData& ns,
     commandBuilder.append("new", returnNew);
     commandBuilder.append("upsert", upsert);
 
+    if (getMaxWireVersion() >= 4) {
```
Comment (amidvidy) [inline src/mongo/client/dbclient.cpp]: if the maxWireVersion is too low, should we uassert that the operation is not supported on the remote server?

File: src/mongo/client/dbclient.cpp (hunk 2 above)
Comment (markbenvenuto) [inline src/mongo/client/dbclient.cpp]: Yes. This code pattern is also used throughout the driver.

File: src/mongo/client/dbclient.cpp (hunk 3 above)
Comment (markbenvenuto) [inline src/mongo/client/dbclient.cpp]: It is not required as part of this change, and I am not sure if there is a relevant wire version to check.

File: src/mongo/integration/standalone/dbclient_test.cpp (hunk 1 above)
Comment (markbenvenuto) [inline src/mongo/integration/standalone/dbclient_test.cpp]: I do not think it is worth the effort. We do not do in any other tests, but I have checked the `system.profile` collection by hand.

**Comment by amidvidy** [comment]
LGTM

File: src/mongo/client/dbclient.cpp (hunk 3 above)
Comment (acmorrow) [inline src/mongo/client/dbclient.cpp]: I feel like ignoring a user requested writeConcern is dangerous. I think we should take the suggestion from @amidvidy

File: src/mongo/integration/standalone/dbclient_test.cpp (hunk 1 above)
Comment (acmorrow) [inline src/mongo/integration/standalone/dbclient_test.cpp]: I agree, not worth the effort.

**Comment by acmorrow** [comment]
LGTM mod uassert per suggestion. No need to re-post.

**Comment by markbenvenuto** [comment]
Merged with https://github.com/mongodb/mongo-cxx-driver/commit/76be500a206bf0d964ae6db81378b22476d9bce0

**Pull request closed without merging**